File: 2번 문제.java
import java.util.*;

public class Solution {
    public static int solution(int sum, int[] coins) {
        List<String> resultArr = new ArrayList<>();
        for (int c : coins) {
            dfs(c, new ArrayList<>(Collections.singletonList(c)), sum, coins, resultArr);
        }
        Set<String> uniqueResultArr = new HashSet<>(resultArr);

        return uniqueResultArr.size();
    }

    private static void dfs(int s, List<Integer> arr, int sum, int[] coins, List<String> resultArr) {
        if (s > sum) {
            return;
        }
        if (s == sum) {
            resultArr.add(arr.stream().sorted().map(String::valueOf).collect(Collectors.joining()));
            return;
        }
        for (int coin : coins) {
            dfs(s + coin, new ArrayList<>(arr) {
                {
                    add(coin);
                }
            }, sum, coins, resultArr);
        }
    }
}

File: 2번 문제.js
function solution (sum, coins){
  let resultArr = [];
  for(let c of coins){
    dfs(c, [c]);
  }
  function dfs(s, arr){
    if(s > sum){
      return;
    }
    if(s === sum){
      resultArr.push(arr.sort().join(''));
      return;
    }
    for(let coin of coins){
      dfs(s + coin, [...arr, coin]);
    }
  }

  return [...new Set(resultArr)].length;
}
